modules/core/src/main/scala/fs2/kafka/TransactionalKafkaProducer.scala

```diff
@@ -42,90 +45,142 @@ abstract class TransactionalKafkaProducer[F[_], K, V] {
-              F.delay(producer.initTransactions())
-            }
-          }
-          .map { withProducer =>
-            new TransactionalKafkaProducer[F, K, V] {
-              override def produce[G[+_], P](
-                records: TransactionalProducerRecords[F, G, K, V, P]
-              ): F[ProducerResult[Chunk, K, V, P]] =
-                produceTransaction(records)
-                  .map(ProducerResult(_, records.passthrough))
-
-              private[this] def produceTransaction[G[+_], P](
-                records: TransactionalProducerRecords[F, G, K, V, P]
-              ): F[Chunk[(ProducerRecord[K, V], RecordMetadata)]] = {
-                if (records.records.isEmpty) F.pure(Chunk.empty)
-                else {
-                  val batch =
-                    CommittableOffsetBatch.fromFoldableMap(records.records)(_.offset)
-
-                  val consumerGroupId =
-                    if (batch.consumerGroupIdsMissing || batch.consumerGroupIds.size != 1)
-                      F.raiseError(ConsumerGroupException(batch.consumerGroupIds))
-                    else F.pure(batch.consumerGroupIds.head)
-
-                  consumerGroupId.flatMap { groupId =>
-                    withProducer { producer =>
-                      F.bracketCase(F.delay(producer.beginTransaction())) { _ =>
-                        val producerRecords: ArrayBuffer[ProducerRecord[K, V]] =
-                          new ArrayBuffer(records.records.size)
-
-                        records.records.foreach { committable =>
-                          committable.foldable.foldLeft(committable.records, ()) {
-                            case (_, record) =>
-                              producerRecords += record
-                              ()
-                          }
-                        }
-
-                        Chunk
-                          .buffer(producerRecords)
-                          .traverse(
-                            KafkaProducer.produceRecord(keySerializer, valueSerializer, producer)
-                          )
-                          .map(_.sequence)
-                          .flatTap { _ =>
-                            F.delay {
-                              producer.sendOffsetsToTransaction(
-                                batch.offsets.asJava,
-                                groupId
-                              )
+  ) {
+
+    type Records = CommittableProducerRecords[F, K, V]
+    type RecordsChunk = Chunk[Records]
+    type RecordGrouper = RecordsChunk => F[Chunk[(KafkaByteProducer, RecordsChunk)]]
+
+    private[this] def resource(
+      buildGrouper: Blocker => Resource[F, RecordGrouper]
+    ): Resource[F, TransactionalKafkaProducer[F, K, V]] =
+      Resource.liftF(settings.keySerializer).flatMap { keySerializer =>
+        Resource.liftF(settings.valueSerializer).flatMap { valueSerializer =>
+          settings.blocker.fold(Blockers.producer)(Resource.pure[F, Blocker]).flatMap { blocker =>
```

Comment:
We should probably introduce a `Blocker.transactionalProducer`, which uses a cached thread-pool rather than a single executor thread pool, since we'll want multiple transactions to run in parallel. `Blockers.producer` will only allow one transaction to run at-a-time.